proto/concept.proto

```diff
@@ -0,0 +1,963 @@
+                Annotations annotation_filter = 2;
+            }
+            // Only one filter can be set at a time. repeated can't be used in a oneof
+            message AttributeTypes {
+                repeated AttributeType attribute_types = 1;
+            }
+            message Annotations {
+                repeated Type.Annotation annotations = 1;
+            }
+        }
+        message ResPart {
+            repeated Attribute attributes = 1;
+        }
+    }
+
+    message SetHas {
+        message Req {
+            Attribute attribute = 1;
+        }
+        message Res {}
+    }
+
+    message UnsetHas {
+        message Req {
+            Attribute attribute = 1;
+        }
+        message Res {}
+    }
+
+    message GetRelations {
+        message Req {
+            repeated RoleType role_types = 1;
+        }
+        message ResPart {
+            repeated Relation relations = 1;
+        }
+    }
+
+    message GetPlaying {
+        message Req {}
+        message ResPart {
+            repeated RoleType role_types = 1;
+        }
+    }
+}
+
+message Entity {
+    bytes iid = 1;
+    EntityType entity_type = 2;
+    bool inferred = 3;
+}
+
+message Relation {
+    bytes iid = 1;
+    RelationType relation_type = 2;
+    bool inferred = 3;
+
+    message AddPlayer {
+        message Req {
+            RoleType role_type = 1;
+            Thing player = 2;
+        }
+        message Res {}
+    }
+
+    message RemovePlayer {
+        message Req {
+            RoleType role_type = 1;
+            Thing player = 2;
+        }
+        message Res {}
+    }
+
+    message GetPlayers {
+        message Req {
+            repeated RoleType role_types = 1;
+        }
+        message ResPart {
+            repeated Thing things = 1;
+        }
+    }
+
+    message GetPlayersByRoleType {
+        message RoleTypeWithPlayer {
+            RoleType role_type = 1;
+            Thing player = 2;
+        }
+        message Req {}
+        message ResPart {
+            repeated RoleTypeWithPlayer role_types_with_players = 1;
+        }
+    }
+
+    message GetRelating {
+        message Req {}
+        message ResPart {
+            repeated RoleType role_types = 1;
+        }
+    }
+}
+
+message Attribute {
+    bytes iid = 1;
+    AttributeType attribute_type = 2;
+    Value value = 3;
+    bool inferred = 4;
+
+    message Value {
+        oneof value {
+            string string = 1;
+            bool boolean = 2;
+            sint64 long = 3;
+            double double = 4;
+            // time since epoch in milliseconds
+            sint64 date_time = 5;
+        }
+    }
+
+    message GetOwners {
+        message Req {
+            optional ThingType filter = 1;
+        }
+        message ResPart {
+            repeated Thing things = 1;
+        }
+    }
+}
+
+message Type {
+    message Req {
+        oneof req {
+            RoleType.Req role_type_req = 1;
+            ThingType.Req thing_type_req = 2;
```

Comment:
Let's make `thing_type` requests come first since they're more 'primary' in our minds than role types (apply throughout)